tests/translator/output/api_with_ip_deny_list.json

```diff
@@ -0,0 +1,206 @@
+            {
+              "__Stage__": "*", 
+              "__ApiId__": {
+                "Ref": "ServerlessRestApi"
+              }
+            }
+          ]
+        }
+      }
+    }, 
+    "ServerlessRestApiProdStage": {
+      "Type": "AWS::ApiGateway::Stage", 
+      "Properties": {
+        "DeploymentId": {
+          "Ref": "ServerlessRestApiDeployment58bae06e3f"
+        }, 
+        "RestApiId": {
+          "Ref": "ServerlessRestApi"
+        }, 
+        "StageName": "Prod"
+      }
+    }, 
+    "ServerlessRestApiDeployment58bae06e3f": {
+      "Type": "AWS::ApiGateway::Deployment", 
+      "Properties": {
+        "RestApiId": {
+          "Ref": "ServerlessRestApi"
+        }, 
+        "Description": "RestApi deployment id: 58bae06e3fc57cd27e4edfbb81d1d5af271892f0", 
+        "StageName": "Stage"
+      }
+    }, 
+    "MyFunctionRole": {
+      "Type": "AWS::IAM::Role", 
+      "Properties": {
+        "ManagedPolicyArns": [
+          "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+        ], 
+        "AssumeRolePolicyDocument": {
+          "Version": "2012-10-17", 
+          "Statement": [
+            {
+              "Action": [
+                "sts:AssumeRole"
+              ], 
+              "Effect": "Allow", 
+              "Principal": {
+                "Service": [
+                  "lambda.amazonaws.com"
+                ]
+              }
+            }
+          ]
+        }
+      }
+    }, 
+    "ServerlessRestApi": {
+      "Type": "AWS::ApiGateway::RestApi", 
+      "Properties": {
+        "Body": {
+          "info": {
+            "version": "1.0", 
+            "title": {
+              "Ref": "AWS::StackName"
+            }
+          }, 
+          "paths": {
+            "/get": {
+              "put": {
+                "x-amazon-apigateway-integration": {
+                  "httpMethod": "POST", 
+                  "type": "aws_proxy", 
+                  "uri": {
+                    "Fn::Sub": "arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${MyFunction.Arn}/invocations"
+                  }
+                }, 
+                "responses": {}
+              }
+            }
+          }, 
+          "swagger": "2.0", 
+          "x-amazon-apigateway-policy": {
+            "Version": "2012-10-17", 
+            "Statement": [
+              {
+                "Action": "execute-api:Invoke", 
+                "Resource": [
+                  {
+                    "Fn::Sub": [
+                      "arn:aws:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/PUT/get", 
+                      {
+                        "__Stage__": "Prod", 
+                        "__ApiId__": "ServerlessRestApi"
+                      }
+                    ]
+                  }
+                ], 
+                "Effect": "Allow", 
+                "Principal": {
+                  "AWS": [
+                    "12345"
```

Comment:
Ok now I see why you included IAM whitelist along with your IP deny... This is pretty interesting. The way API GW resource policies work is pretty confusing when you're hand-editing the aspen policy, but the SAM shorthand syntax used in certain combinations could lead customers to think they've secured their API when they've really opened it up to any IAM user. For example, any resource policy that adds Allow Principal: "*", but mixes that with IAM auth of specific accounts could be problematic, right (I'm actually not sure what the API GW behavior is in this case)? We'll need to discuss this with security. SAM may need to do some checking and disallow certain combinations of the shorthand resource policy types.

Reply:
Thanks for the valuable feedback @jlhood! Working my way through these...should update soon!

Reply:
> Ok now I see why you included IAM whitelist along with your IP deny... This is pretty interesting. The way API GW resource policies work is pretty confusing when you're hand-editing the aspen policy, but the SAM shorthand syntax used in certain combinations could lead customers to think they've secured their API when they've really opened it up to any IAM user. For example, any resource policy that adds Allow Principal: "*", but mixes that with IAM auth of specific accounts could be problematic, right (I'm actually not sure what the API GW behavior is in this case)? We'll need to discuss this with security. SAM may need to do some checking and disallow certain combinations of the shorthand resource policy types.

I just checked the behavior on the console - It throws an error "Invalid policy document. Please check the policy syntax and ensure that Principals are valid." When my policy statement looks like this - 
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": [
                    "arn:aws:iam::{{otherAWSAccountID}}:root",
                    "arn:aws:iam::{{otherAWSAccountID}}:user/{{otherAWSUserName}}",
                    "arn:aws:iam::{{otherAWSAccountID}}:role/{{otherAWSRoleName}}"
                ]
            },
            "Action": "execute-api:Invoke",
            "Resource": [
                "execute-api:/{{stageNameOrWildcard*}}/{{httpVerbOrWildcard*}}/{{resourcePathOrWildcard*}}"
            ]
        },
        {
            "Effect": "Deny",
            "Principal": "*",
            "Action": "execute-api:Invoke",
            "Resource": "execute-api:/{{stageNameOrWildcard}}/{{httpVerbOrWildcard}}/{{resourcePathOrWildcard}}",
            "Condition" : {
                "IpAddress": {
                    "aws:SourceIp": [ "{{sourceIpOrCIDRBlock}}", "{{sourceIpOrCIDRBlock}}" ]
                }
            }
        },
        {
            "Effect": "Allow",
            "Principal": "*",
            "Action": "execute-api:Invoke",
            "Resource": "execute-api:/{{stageNameOrWildcard}}/{{httpVerbOrWildcard}}/{{resourcePathOrWildcard}}"
        }

    ]
}
```

Given this, do we validate on SAM or do you think that this error is sufficient?

Reply:
Which principal did it say was invalid?

Reply:
Ok great. If API GW's control plane is doing the error checking, I don't think we need to add extra error checking in the SAM layer. Thanks for verifying that!